### PR TITLE
Fixed previous weight handling for DCASGD optimizer.

### PR DESCRIPTION
**Pull request by sergeykolychev**
While porting latest Python changes to the Perl interface I stumbled upon something that looks like a small logic error.

### CODE REVIEW

**Comment by piiswrong** [comment]
Without a copy weight previous will always be the same with weight. It should have been a state

**Comment by sergeykolychev** [comment]
The logic still does not compute (I may be missing something)
weight - previous_weight is always 0
Can you devise what is should be from ttps://arxiv.org/pdf/1609.08326.pdf page 5
I do not see any async about current code, do you have time to check this ?

**Comment by sergeykolychev** [comment]
I do not know if it actually implements the paper (I do not know about inner workings of mxnet that deep yet), but the logic now computes in my mind.

File: python/mxnet/optimizer.py [hunk 1]
```diff
@@ -349,26 +351,12 @@ def update(self, index, weight, grad, state):
...
-        else:
-            assert self.momentum == 0.0
-            if self.weight_previous.has_key(index):
-                weight[:] += -lr * (grad + wd * weight + self.lamda \
-                                    * grad * grad * (weight - self.weight_previous[index]))
-                self.weight_previous[index] = weight
-            else:
-                weight[:] += -lr * (grad + wd * weight)
-                self.weight_previous[index] = weight
+        mom, previous_weight = state
```
Comment (piiswrong) [inline python/mxnet/optimizer.py]: This wastes memory and compute. better use different states and test for momentum == 0

**Comment by sergeykolychev** [comment]
@piiswrong 
Eric, could you please review the latest change and see if it makes sense to you.

**Pull request merged**